Merge to Dev: Story 609 - add dynamic page functionality for Static Pages

harryface opened the pull request:
Allows dynamic static pages to be created on the fly with a dynamic URL too

- Levergaed middleware
- No model change
- View tests included
- Former static page URL modified (urls to existing static pages tested and still working as expected)

NB: If the url clashes with a hard-coded url, hard-coded would take precedence.

tompollard commented:
Thanks @harryface, this will be useful! 

We have just merged https://github.com/MIT-LCP/physionet-build/pull/1695 which adds testing functionality to URLs. Please could you rebase to deal with conflicts and then take a look at the failing tests?

Most likely there will need to be some additional tweaks to urls.py to address the tests added in https://github.com/MIT-LCP/physionet-build/pull/1695. We can chat about this tomorrow if anything isn't clear.

harryface commented:
Rebased and refactored.
The `test_url` assumes that every url that receives a positional argument should also be called without one, this one does not and as a result fails...

bemoody commented:
I think we want to restrict these pages to the /about/ hierarchy or other well-defined hierarchies.  I *don't* want to have new pages created all over the place that we then have to worry about conflicting with.

> The test_url assumes that every url that receives a positional argument should also be called without one, this one does not and as a result fails...

`test_url` doesn't make any such assumptions.  The `views.static_view` function is invoked both with and without an argument because there are two separate `path()`s for it (one that has an argument and one that doesn't.)

harryface commented:
> I think we want to restrict these pages to the /about/ hierarchy or other well-defined hierarchies. I _don't_ want to have new pages created all over the place that we then have to worry about conflicting with.
> 
> > The test_url assumes that every url that receives a positional argument should also be called without one, this one does not and as a result fails...
> 
> `test_url` doesn't make any such assumptions. The `views.static_view` function is invoked both with and without an argument because there are two separate `path()`s for it (one that has an argument and one that doesn't.)

With the rebase, I forgot to remove the about URL, I was wrong.
Should I go ahead and restrict it to about or still allow it as a catch all?

tompollard commented:
> Should I go ahead and restrict it to about or still allow it as a catch all?

Yeah, I think this is a good idea.

alistairewj commented:
Agreed!

bemoody commented:
>> Should I go ahead and restrict it to about or still allow it as a catch all?

> Yeah, I think this is a good idea.

Which one?

tompollard commented:
> > > Should I go ahead and restrict it to about or still allow it as a catch all?
> 
> > Yeah, I think this is a good idea.
> 
> Which one?

Whoops, the first one!

tompollard commented on physionet-django/console/forms.py:
```diff
@@ -760,6 +760,54 @@ def save(self):
+        regex=r"^[-\w/\.~]+$",
+        help_text=(
+            "Url should be unique. If it clashes with a static url, "
+            "The static url will take precedence."
```
Perhaps tweak slightly to:

"URL should be unique. If the new URL clashes with a static url, "
"the static url will take precedence."

Ideally, I think we should also 
- include an example of the kind of URL path that we are expecting (e.g. `/about/path`)
- specify that the path must begin with `about`.

tompollard commented on physionet-django/console/forms.py:
```diff
@@ -760,6 +760,54 @@ def save(self):
+            same_url = same_url.exclude(pk=self.instance.pk)
+
+        if same_url.exists():
+            raise forms.ValidationError(f"Static page with url {url} already exists")
```
Tweak to:

`Static page with URL {url} already exists.`

tompollard commented on physionet-django/console/views.py:
```diff
@@ -2363,6 +2363,54 @@ def static_pages(request):
+        static_page_form = forms.StaticPageForm(instance=static_page, data=request.POST)
+        if static_page_form.is_valid():
+            static_page_form.save()
+            messages.success(request, "The static page was successfully editted.")
```
typo ("editted")

tompollard commented on physionet-django/console/forms.py:
```diff
@@ -760,6 +760,54 @@ def save(self):
+        label="URL",
+        max_length=100,
+        regex=r"^[-\w/\.~]+$",
+        help_text=(
```
If possible, it would be good if the help text was also displayed on the "edit" page. Currently I only see it on the creation page.

tompollard reviewed (commented):
Thanks @harryface, this is looking good to me. I've annotated the code with a few comments, and also listed a few thoughts below:

- Do any of the fixtures need updating? i.e. are the pages/paths listed in https://github.com/MIT-LCP/physionet-build/blob/dev/physionet-django/physionet/fixtures/demo-physionet.json correct?
- Currently we have an about page at https://physionet.org/about/. Will this path be tricky to deal with or can allow this by simply supporting an empty string after "/about/"?

I assume you are working on the changes so that paths are restricted to `/about/`. I'll do some testing once this is pushed.

tompollard commented:
Looks good to me, thanks @harryface!
